### PR TITLE
MBTA Demo

### PR DESCRIPTION
**Pull request by wangandi**
Feel free to change reviewers to whomever is in charge of integrating new demos these days.

How do I test that buildkite will successfully build the actual docker containers corresponding to the demo?

Currently, the version of the demo in this PR should be sufficient for somebody to download and try out. Turning this into an automated test is hampered by my current inability to use docker compose secrets to pass an api key in.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4225)

### CODE REVIEW

**Comment by quodlibetor** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 

> How do I test that buildkite will successfully build the actual docker containers corresponding to the demo?

mzbuild does this automatically, if it doesn't work it is a bug, but it _should_ work the same in buildkite as it does locally.

> Turning this into an automated test is hampered by my current inability to use docker compose secrets to pass an api key in.

I'm happy to dig in with you on getting the secrets in after I've reviewed it. It would be nice to make sure that this works before merging so that we're assured that the demo is always working.

**Comment by wangandi** [comment]
Automated build has failed with
```

docker push materialize/mbta-demo:mzbuild-PYFJVSPRCK6OOSKIZDIBY5DQBEXAOED2
--
  | The push refers to repository [docker.io/materialize/mbta-demo]
  | 9dd356fd0f88: Preparing
  | 54ad9f2f3aeb: Preparing
  | 7bf9f3d72448: Preparing
  | 2db04a1cd49d: Preparing
  | 28ba7458d04b: Preparing
  | 838a37a24627: Waiting
  | a6ebef4a95c3: Waiting
  | b7f7d2967507: Waiting
  | denied: requested access to the resource is denied
  | Traceback (most recent call last):
  | File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
  | "__main__", mod_spec)
  | File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
  | exec(code, run_globals)
  | File "/var/lib/buildkite-agent/builds/buildkite-builders-i-0dddfa16c04d447d3-1/materialize/tests/ci/test/build.py", line 116, in <module>
  | main()
  | File "/var/lib/buildkite-agent/builds/buildkite-builders-i-0dddfa16c04d447d3-1/materialize/tests/ci/test/build.py", line 39, in main
  | d.push()
  | File "/var/lib/buildkite-agent/builds/buildkite-builders-i-0dddfa16c04d447d3-1/materialize/tests/misc/python/materialize/mzbuild.py", line 489, in push
  | spawn.runv(["docker", "push", self.spec()])
  | File "/var/lib/buildkite-agent/builds/buildkite-builders-i-0dddfa16c04d447d3-1/materialize/tests/misc/python/materialize/spawn.py", line 66, in runv
  | args, cwd=cwd, stdin=stdin, stdout=stdout, stderr=stderr, check=True, env=env,
  | File "/usr/lib/python3.6/subprocess.py", line 438, in run
  | output=stdout, stderr=stderr)
  | subprocess.CalledProcessError: Command '['docker', 'push', 'materialize/mbta-demo:mzbuild-PYFJVSPRCK6OOSKIZDIBY5DQBEXAOED2']' returned non-zero exit status 1.

```

What does this mean?

File: demo/mbtaupsert/src/main.rs
Comment (benesch) [inline demo/mbtaupsert/src/main.rs:279]: The control flow in this function is really hard to follow due to the nesting.

File: demo/mbtaupsert/src/main.rs
Comment (benesch) [inline demo/mbtaupsert/src/main.rs:284]: This violates standard Unix conventions for error handling. The error should be printed to stderr, not stdout, and the return code of the program should be nonzero.

If you don't exit with a failing status code, then CI won't be able to check whether this demo succeeded or not.

File: demo/mbtaupsert/src/kafka_interaction.rs
Comment (benesch) [inline demo/mbtaupsert/src/kafka_interaction.rs:297]: This is way too much code to duplicate. We're trying to factor these methods into the `kafka-util` crate. The create topics function should already do what you need. We should figure out something for the message production too, because hacking up schema IDs here strikes me as really sketchy, and duplicates a lot of logic that already exists in `interchange`.

**Review by benesch — CHANGES_REQUESTED** [review]
This is a really awesome demo, but I'm not sure we're prepared to support this as an eng org. We're barely keeping the lights on with the demos that we have already, e.g. https://github.com/MaterializeInc/materialize/issues/4223. This demo is particularly scary from a maintenance perspective because it relies heavily on the MBTA data set. That's a big part of what makes it so cool, but it's definitely an extra challenge.

I may be unfairly picking on adding demos, but we've accumulated a ton of demo cruft over the past year and I'm leery of adding another demo without pruning what we have. For example the `trx` demo is mostly unused and broken and duplicates all of the code in the `chbench` demo: https://github.com/MaterializeInc/materialize/tree/main/demo/trx. It's particularly bad that every demo reinvents the wheel wrt creating topics and writing to Kafka.

I'm curious to hear what other folks think. Part of what scares me is that the expectation of demos in the main repo seems to have have evolved from from "cool thing that maybe works; you may need to modify/debug it!" to "P0/P1 demo doesn't work out of the box". 

One answer may be that it's time to delete/conslidate demos that aren't pulling their weight. Are `simple` and `http_logs` still relevant now that we have all of these bigger demos? Can they be merged in with some of the other demos? At the very very least, if this goes in, we should not add any more demos without deleting/consolidating some of the existing demos.

> I'm happy to dig in with you on getting the secrets in after I've reviewed it. It would be nice to make sure that this works before merging so that we're assured that the demo is always working.

CI must absolutely not depend on fetching data from the MBTA. Non-determinism in CI is the root of all evil. We accept just a little bit of it where we have to (fetching APT packages and images from Docker Hub), but blocking merging PRs on MBTA flakiness is a recipe for frustration.

**Comment by wangandi** [comment]
> CI must absolutely not depend on fetching data from the MBTA. Non-determinism in CI is the root of all evil. We accept just a little bit of it where we have to (fetching APT packages and images from Docker Hub), but blocking merging PRs on MBTA flakiness is a recipe for frustration.

The fetching data from the MBTA is intended for a load test-style automated workflow because it would be nice to run some upsert load tests and load tests on real data. 

But you have reminded me that we generally have, amongst the merge checks, a sanity check for each demo. Would it be acceptable for the sanity check to replay a fixed archive of <insert time here worth of> of MBTA data instead of fetching anything? (replay functionality is already in this PR) Though I don't know what would be the best way to make that archive available to automation, or what our limit on the size of such an archive would be.

**Comment by wangandi** [comment]
> I'm curious to hear what other folks think. Part of what scares me is that the expectation of demos in the main repo seems to have have evolved from from "cool thing that maybe works; you may need to modify/debug it!" to "P0/P1 demo doesn't work out of the box".

Pinging @awang for opinions on demo maintenance expectations.

I'm going to try to resolve the code issues that @benesch brought up ASAP, but this comment brings up some interesting questions of marketing vs. engineering maintenance trade-offs. 

1. I feel that we should have a standard policy for having WIP demos in the main repository. Maintaining this in MaterializeInc/mbtaupsert is getting increasingly expensive for me because stuff like resolving dependencies on subpackages in this repository, the lack of access to the tooling in this repository, the lack of testing support meaning that I have to manually retest everything every time I change something or update to the latest Materialize change, etc. Personally, I think having more demos in the main repository instead of less would help with the "every demo reinvents the wheel wrt creating topics and writing to Kafka" problem because 1) we would have greater visibility into common code between demos. 2) there would be more than one pair of eyes on each demo.

2.  My perception of the best thing to do re: marketing is to 1) merge this to main 2) publish a blog post about this. That way when people see the blog post, they can immediately download the demo and see how cool we are. For future demos, would it be preferable to just publish the blog post as fast as possible and leave readers to engineer how they can make the demo happen themselves? At least until we have a sales engineer on board?

**Comment by awang** [comment]
Demos that are linked in to by our docs (ie http://materialize.io/*) or blog posts should be treated as being in "production".  It's not as big a deal now, but as we direct more traffic to Materialize via marketing campaigns it will matter more.

I have less of an opinion on demos that aren't linked to, such as `trx`

Re. publishing timing (#2) the ideal is to have a fully-fleshed blog post that readers can try themselves, but I'm also pragmatic -- there are time+effort tradeoffs, as well as diminishing returns. We can also discuss off GH about other ways to structure the blog post (like splitting it into multiple parts)

**Comment by cuongdo** [comment]
@benesch What amount of sanity checking would make you feel comfortable with merging this demo into `materialize`? Would Andi's suggestion of replaying pre-downloaded MBTA data (e.g. retrieved from a Materialize S3 bucket) address your concerns?

File: demo/mbtaupsert/src/main.rs
Comment (wangandi) [reply demo/mbtaupsert/src/main.rs:377]: I have tried adding more comments.

File: demo/mbtaupsert/src/kafka_interaction.rs
Comment (wangandi) [reply demo/mbtaupsert/src/kafka_interaction.rs]: I just commented out BYO support wholesale since it is not necessary for a user to try out.

**Comment by benesch** [comment]
> What amount of sanity checking would make you feel comfortable with merging this demo into materialize? 

I think the most important thing is a commitment that fixing this demo if it breaks is not a priority. I would really like to avoid another situation like #4223, where the demo breaking is considered a P0/P1 event. I don't necessarily disagree that we should be holding a few user-facing demos that are covered in the doc to that standard, but the demos that _are_ held to that standard need to be production quality code with good test coverage, and ideally those demos would be _just_ demos, not combination demos/integration tests/benchmarking harnesses.

**Comment by wangandi** [comment]
A more immediate question: given the fact that we don't have a commitment to fix anything until the demo is publicized via blog post or docs, is it acceptable to merge this PR now (provided the issues raised in the existing code have satisfactorily resolved)? And then I can submit some follow up PRs for sanity checking?

> ideally those demos would be just demos, not combination demos/integration tests/benchmarking harnesses.

The downside of having demos just being demos is that it would be mean we'd unnecessarily reengineer the data ingest pipeline every time we want to run a one-off test with a nontrivial amount of data under new conditions.

Would it be sufficient to put a disclaimer that we only support, say, the automatic docker setup, and we are not liable if the user has some custom kafka setup + wants to run with custom kafka properties?

Or would it preferable to have two versions of the demo:
1. a minimalistic version with only features that we can fully commit to supporting
2. one with the demo + options that make it easier to extend for testing purposes.

**Review by benesch — COMMENTED** [review]
Sorry for the slow review. There's quite a bit of code here! This is mentioned in the PR guide, but a PR of this size is hard to review. The diffstat is much more sane without the Cargo.lock file, but it's still definitely on the heftier size. I know it's hard to split up something like this where you're importing a pre-existing project—and the copious docs are very much appreciated—so maybe just a bit more warning that this is coming down the pipe would be helpful in the future. (Alternatively, if we make it more clear what demos are WIP and what demos are production ready, _starting_ the demo in this repo from scratch could work too.)

It seems to me that you've made life hard on yourself by splitting the logic between the Bash script and the Rust program. For example config file parsing is implemented twice, once in awk, and once in Rust. I think it would be a good bit simpler and less error prone to have all of that logic in Rust, and to have that logic use the `csv` crate rather than rolling your own CSV parser? 

Similarly, streaming the server-sent events seems to be a rather intricate dance between backgrounded curl commands and a hand-rolled tailing file reader that is rather complicated because it doesn't always read a complete line. Did you consider using something like https://docs.rs/eventsource/0.5.0/eventsource/ instead? 

I'm still unconvinced that we're prepared to maintain this complexity long term. If in six months or a year, the demo has rotted—say the MBTA dataset has changed fundamentally and they've deprecated the old APIs—are we comfortable deleting it even though that will rot the blog post too? If the answer to this question is "yes", then I think it's ok to merge this in, and we can see how the maintenance feels over time. If it's too much, we can hit the undo switch. But if the answer is "no", we must figure out how to get the demo working again, then I'm nervous that the blog post will have essentially locked in a maintenance burden beyond what we anticipated.

To be clear, I want the answer to the above to be "no." If you're comfortable with that  and you get a firm commitment from @awang that this demo does not need to be supported, then these are the only things I care strongly about:

  * Removing commented-out code. It's really hard to tell when commented-out code is a mistake that someone forgot to uncomment or whether it's a feature that is temporarily disabled. Even if it's the latter, since `cargo check` can't see commented code, that code tends to rot by the time someone goes to re-enable it anyway. I'd much prefer to either leave that code in, even though it's unused, or delete it entirely.

 * Adding `set -e` to the Bash script to better protect against errors.

 * Fixing the documentation links.

Long term, however, I'd love to see a removal of the Bash script entirely. We've been trying to move away from Bash, since it's error prone and doesn't scale well. (For example, chbench migrated from a monstrous Bash script to the much tamer mzconduct workflows.) In many cases Python has been the right choice, although I think in this case the complexity could be reduced by approximately half by merging the Bash script into the mbtaupsert Rust binary and using a carefully chosen third-party crate or two. I'm definitely up for helping on this, if I wouldn't be step-in non your toes.

> Would it be sufficient to put a disclaimer that we only support, say, the automatic docker setup, and we are not liable if the user has some custom kafka setup + wants to run with custom kafka properties?

Even the automatic Docker setup has a way of being flaky, unfortunately. With as many moving parts as are required to run Materialize, it's very easy for something to go wrong, even with Docker.

> Reviewed 5 of 22 files at r1, 4 of 12 files at r3, 3 of 11 files at r4, 8 of 10 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4225#-:-MHtgHh54lXVO9EbBPau:b13qh00)* status: all files reviewed, 12 unresolved discussions (waiting on @quodlibetor and @wangandi)

___
*[Cargo.toml, line 4 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHqU-DnECclcwEcZfrB:-MHqU-Do-yIUH7b0Romf:bd3eztc) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/Cargo.toml#L4)):*
> ```TOML
> members = [
>     "demo/billing",
>     "demo/mbtaupsert",
> ```

Can we just call this `mbta` or `transit`? The "upsert" bit is kind of confusing to users, I think? The headline feature is that it's a super cool transit demo, I think! The fact that it uses "upsert" internally I think is a lot less relevant—it's pretty hard to even see the upserts getting used because of the way the upserts happen right in the middle of the transformation pipeline. The "transit in server-sent events format into real-time Materialize" is what I see at a glance, and what I think most users will see too.
___
*[bin/lint, line 77 at r2](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHJVMAaoxLxFDqQJJpf-r2-77:-MHqU7FM0_doZiQ9Y3Zk:b-ucfknh) ([raw file](https://github.com/materializeinc/materialize/blob/643a0ab27355696ca7cedb121a19e025ee6aa91e/bin/lint#L77)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ignoring this entirely is too broad. The demo files should get copyright headers. If that causes problems for the data files, let's mark those as CSVs with `.csv` extensions and add `csv` to the regex on L72.
</blockquote></details>

Thanks!
___
*[demo/mbtaupsert/mzcompose.yml, line 45 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHtn_rP-l-a-8CkSqSl:-MHtn_rP-l-a-8CkSqSm:b-cntvwm) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/mzcompose.yml#L45)):*
> ```YAML
>       - ./archive:/workdir/archive
>       - ./demo:/workdir/configs
>       - ./workspace:/workdir/workspace
> ```

See below, but I think you might as well just mount `.` to `/workdir`.
___
*[demo/mbtaupsert/mzcompose.yml, line 68 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHtck1eAovPzWqY2q28:-MHtck1eAovPzWqY2q29:b-myevkt) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/mzcompose.yml#L68)):*
> ```YAML
>     #  - step: run
>     #    service: peeker
>     #    command: --only-initialize --queries q01
> ```

I think for now we should just delete all the CI verification and secret passing stuff, and just say that if the demo breaks, that's ok.
___
*[demo/mbtaupsert/mzcompose.yml, line 70 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHtcnEXAjTdEfLrNJ77:-MHtcnEXAjTdEfLrNJ78:b-ux1kzm) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/mzcompose.yml#L70)):*
> ```YAML
>     #    command: --only-initialize --queries q01
> 
>     # Helper workflows
> ```

These don't look like helper workflows. They look like actual user facing workflows to me. It would probably be a good idea to make a `bring-up` helper workflow that starts everything but the `meta-demo` service that's called by `start-live-data` and `replay`, so that you don't have to duplicate that stuff between the two workflows.
___
*[demo/mbtaupsert/README.md, line 26 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHte36VDoAiiXzcRUQc:-MHte36VDoAiiXzcRUQd:bgq73fr) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/README.md#L26)):*
> ```Markdown
> You will be able to issue SQL queries and maintain views on the live state of
> the Boston-area public transit system.
> [doc/mbta-predictions.md](/doc/mbta-predictions.md) contains some inspiration to
> ```

This link seems broken, and I'm not sure where it's supposed to point, either!
___
*[demo/mbtaupsert/README.md, line 73 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHtdIQ73olL-9bb1K1y:-MHtdIQ73olL-9bb1K1z:bv0m1ny) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/README.md#L73)):*
> ```Markdown
> and then produce a key-value Kafka topic out of it.
> 
> Look in [doc/mbtaupsert-doc.md](doc/mbtaupsert-doc.md) for more information.
> ```

I think there is a missing link in this file to doc/mbta-setup.md, which seems to have the actual explanation of how to start this up?
___
*[demo/mbtaupsert/ci/Dockerfile, line 21 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHtdcnu0m536aoRUYKg:-MHtdcnu0m536aoRUYKh:bleyx3b) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/ci/Dockerfile#L21)):*
> ```Dockerfile
> #create workspace
> WORKDIR /workdir
> VOLUME [ "/workdir/archive", "/workdir/configs", "/workdir/workspace"]
> ```

This doesn't seem necessary to me. I would just bind mount the entire demo/mbta directory into the container. At the very least you don't need this `VOLUME` directive for bind mounts.
___
*[demo/mbtaupsert/ci/mbta-demo-cmd.sh, line 328 at r3](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHqqkxJ-6GtvWdhXuqx:-MHqqkxJ-6GtvWdhXuqy:b-5gr29x) ([raw file](https://github.com/materializeinc/materialize/blob/81d564acf5fad4b8616a19ecbdf0a2c676479e80/demo/mbtaupsert/ci/mbta-demo-cmd.sh#L328)):*
> ```Shell
>              upon receiving a SIGTERM."
>     exit 1
> esac
> ```

This script is rather dangerous because it doesn't use `set -e`. (The standard idiom is `set -euo pipefail`.) Without at least `set -e`, commands will silently fail, and and the script will progress onwards, probably resulting in a very confusing experience for the user.
___
*[demo/mbtaupsert/src/kafka_interaction.rs, line 297 at r2](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHIg1zxe7sLcqPrYEfQ-r2-297:-MHtf1Ad1jlRZOU_sTcx:bdx8s79) ([raw file](https://github.com/materializeinc/materialize/blob/643a0ab27355696ca7cedb121a19e025ee6aa91e/demo/mbtaupsert/src/kafka_interaction.rs#L297)):*
<details><summary><i>Previously, wangandi (Andi Wang) wrote…</i></summary><blockquote>

I just commented out BYO support wholesale since it is not necessary for a user to try out.
</blockquote></details>

Hmm. I don't think that leaving commented out code strewn about is good practice, as a rule.
___
*[demo/mbtaupsert/src/main.rs, line 279 at r2](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHIg1zxe7sLcqPrYEfR-r2-279:-MHtg2Zt3jyiko04FamX:b-t1r1sw) ([raw file](https://github.com/materializeinc/materialize/blob/643a0ab27355696ca7cedb121a19e025ee6aa91e/demo/mbtaupsert/src/main.rs#L279)):*
<details><summary><i>Previously, wangandi (Andi Wang) wrote…</i></summary><blockquote>

I have tried adding more comments.
</blockquote></details>

Thanks, that helps a bit. It would still be best to try to reduce the nesting with a helper function, I think.
___
*[demo/mbtaupsert/src/main.rs, line 28 at r6](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHthx0W-z0W6Jv3QuPt:-MHthx0X8vWNO38p0ovP:b-u7e9lt) ([raw file](https://github.com/materializeinc/materialize/blob/ecfed12997777d067a778ef2eaa554e1d85f086f/demo/mbtaupsert/src/main.rs#L28)):*
> ```Rust
> extern crate json;
> extern crate rdkafka;
> extern crate reqwest;
> ```

`extern crate` isn't necessary in the 2018 edition.

**Comment by benesch** [comment]
Oh, and long term it would be great to get a basic smoke test running in CI based on archived data, but I think we should skip that and just message this demo as WIP for now.

**Comment by awang** [comment]
> you get a firm commitment from @awang that this demo does not need to be supported


That's generally fine with me, since demos are at the "edge" and not the "core" (ie not foundational, where other parts depend on it). Demo or example code rot is tech debt that I think we'll have to be comfortable with for the short term, in favor of expediency.

**Comment by wangandi** [comment]
> Similarly, streaming the server-sent events seems to be a rather intricate dance between backgrounded curl commands and a hand-rolled tailing file reader that is rather complicated because it doesn't always read a complete line. Did you consider using something like https://docs.rs/eventsource/0.5.0/eventsource/ instead?

It was the first thing I tried months and months ago. The package you linked does not support adding headers, and I needed an API key header. I also tried https://github.com/launchdarkly/rust-eventsource-client (and another package, I think, which I can't find again) for several hours, but I could not get it working with the API key, and I wasn't sure if I should rely on something that is described as "Early stage release for feedback purposes. May contain bugs or performance issues. API subject to change."

That said, upon re-googling for Rust SSE packages, there's a new SSE package as of late June https://crates.io/crates/eventsource-stream, which I consider to be promising because it does not reinvent the HTTP client; it just parses the bytes_stream into the proper format. I can look into that.

**Comment by benesch** [comment]
> It was the first thing I tried months and months ago. The package you linked does not support adding headers, and I needed an API key header.

It does if you pass in a reqwest client whose default headers includes your API key. Here's an example:

https://github.com/benesch/sse-play/blob/66936b7c3a29ad8d0427df078f158ee4ebe58ba8/src/main.rs#L12-L26

**Review by wangandi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4225)* status: 0 of 19 files reviewed, 12 unresolved discussions (waiting on @benesch, @quodlibetor, and @wangandi)

___
*[demo/mbtaupsert/ci/mbta-demo-cmd.sh, line 328 at r3](https://reviewable.io/reviews/materializeinc/materialize/4225#-MHqqkxJ-6GtvWdhXuqx:-MIeoUe5D4ZlgGocdc2X:b-qdtczz) ([raw file](https://github.com/materializeinc/materialize/blob/81d564acf5fad4b8616a19ecbdf0a2c676479e80/demo/mbtaupsert/ci/mbta-demo-cmd.sh#L328)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This script is rather dangerous because it doesn't use `set -e`. (The standard idiom is `set -euo pipefail`.) Without at least `set -e`, commands will silently fail, and and the script will progress onwards, probably resulting in a very confusing experience for the user.
</blockquote></details>

I put in `set -euo pipefail`, but I disabled it in the subshell where the `curl` command runs from because the `set -e` was causing the script to exit out of the indefinite loop and not reconnect to the  stream.

**Review by benesch — APPROVED** [review]
> Reviewed 7 of 21 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4225)* status: 8 of 19 files reviewed, 8 unresolved discussions (waiting on @benesch, @quodlibetor, and @wangandi)

___
*[play/mbta/Cargo.toml, line 2 at r10](https://reviewable.io/reviews/materializeinc/materialize/4225#-MIztaFT4VTjbDQpUEZU:-MIztaFUDDTSE3kZyxi8:b-bicaiq) ([raw file](https://github.com/materializeinc/materialize/blob/157ecf433d023a651eebc3597ade5b9384c07511/play/mbta/Cargo.toml#L2)):*
> ```TOML
> [package]
> name = "mbta_to_mtrlz"
> ```

By convention packages names (as opposed to crate names) should use hyphens, not underscores.

**Pull request merged**
